### PR TITLE
Search and launch apps on windows and linux

### PR DESCRIPTION
**Pull request by KELiON**
It is a simple solution, that just do a `glob` search. 

1) On windows we are searching for .exe or .lnk files in 3 [directories](https://github.com/KELiON/cerebro/compare/feature/windows-apps?expand=1#diff-9febaa8a20fe54d121024454fcaf13d9R7)

2) On Linux we are searching for .desktop files in each directory from `$XDG_DATA_DIRS` + `/applications/`.

Test builds:

https://www.dropbox.com/s/wkc1lncjtx00rpy/Cerebro%20Setup%200.2.3.exe?dl=0
https://www.dropbox.com/s/7y7hrn0epfjanjm/cerebro_0.2.3_amd64.deb?dl=0
https://www.dropbox.com/s/lu5r4k6b18xwzev/cerebro-0.2.3-x86_64.AppImage?dl=0

### CODE REVIEW

File: app/main/plugins/core/basic-apps/linux.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    ...appDirs,
+    ...process.env.XDG_DATA_DIRS.split(':')
+  ]
+}
+
+appDirs = appDirs.map(dir => path.join(dir, 'applications'))
+
+export const DIRECTORIES = uniq([
+  ...appDirs,
+  path.join('usr', 'share', 'app-install', 'desktop')
```
Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: @KELiON we also need `/usr/share/applications` and `/usr/local/share/applications` here

**Comment by maximbaz** [comment]
It looks like the search is matching file names (in addition or instead of `Name` field inside the *.desktop file), this produces a lot of noise among search results.

For example, as I learned now, Chrome generated a bunch of files matching `~/.local/share/applications/*chrome*.desktop`. 

Searching for "Chrome" in Cerebro therefore finds them all, even though they don't even contain a word "Chrome" in their `Name` field.

cerebro_285


Albert launcher doesn't match that, when I search "Chrome" in albert, I get only one result, the correct one.

Plus, as I commented elsewhere, most of my apps have their *.desktop files in `/usr/share/applications` folder, we've got to search that folder too :)

**Comment by KELiON** [comment]
@maximbaz ok, I'll fix these duplications. Can you check what do you have in `$XDG_DATA_DIRS` env variable by `echo $XDG_DATA_DIRS`?

**Comment by maximbaz** [comment]
Thanks! Both $XDG_DATA_HOME and $XDG_DATA_DIRS are not set on my machine.

**Comment by KELiON** [comment]
@maximbaz I've updated file above. Can you check it now? Should be less noise and no duplicates

**Comment by maximbaz** [comment]
Much less noise, yay. However it looks like the app is still not looking inside of `/usr/share/applications`, all results that I have are from `~/.local/share/applications`.

I'm trying to find a way to debug this, to see for example what are the final values of `appDirs`...

File: app/main/plugins/core/basic-apps/linux.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import uniq from 'lodash/uniq'
+import { shellCommand } from 'cerebro-tools'
+
+let appDirs = [
+  path.join(remote.app.getPath('home'), '.local', 'share'),
+  path.join('usr', 'share'),
+  path.join('usr', 'share', 'ubuntu'),
+  path.join('usr', 'share', 'gnome'),
+  path.join('usr', 'local', 'share'),
+  path.join('var', 'lib', 'snapd', 'desktop')
```
Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: these guys ^^ don't have the initial slash, they look like "usr/share" and not like "/usr/share" - and therefore are filtered afterwards as not existing directories.

Comment (KELiON) [inline app/main/plugins/core/basic-apps/linux.js]: Do you have set up environment? If so you can just run application by `npm run dev` and add console.log/debug in this file and check what is inside yours appDirs/DIRECTORIES

Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: unfortunately I don't, but I managed to do it the hard way, through dev tools :)

```
appDirs: 
["/home/maximbaz/.local/share/applications", "usr/share/applications", "usr/share/ubuntu/applications", "usr/share/gnome/applications", "usr/local/share/applications", "var/lib/snapd/desktop/applications", "/tmp/.mount_IKkzVw/usr/share/applications", "share/applications", "applications"]

DIRECTORIES:
["/home/maximbaz/.local/share/applications"]
```

I apply this command in debugger just before DIRECTORIES are set:
`appDirs = appDirs.map(dir => dir[0] == '/' ? dir : ('/' + dir));`

Then appDirs and DIRECTORIES are fixed:

```
appDirs = ["/home/maximbaz/.local/share/applications", "/usr/share/applications", "/usr/share/ubuntu/applications", "/usr/share/gnome/applications", "/usr/local/share/applications", "/var/lib/snapd/desktop/applications", "/tmp/.mount_IKkzVw/usr/share/applications", "/share/applications", "/applications"]

DIRECTORIES = ["/home/maximbaz/.local/share/applications", "/usr/share/applications"]
```

Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: But even then, even if I remove apps_cache.json, the apps from `/usr/share/applications` are missing...

I'll try to get a proper environment and have a closer look.

Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: Yes, that worked, just ensure the initial slash exists, and the `/usr/share/applications` (and others) are taken into account.

**Comment by maximbaz** [comment]
Let me just do more testing and post some results here, later we investigate.

I type "Thunderbird" and I see two results:

cerebro_286

What is the first one, where does it come from? In fact, none of these two actually start the app, they just do nothing. I can start other apps, so the problem is definitely Thunderbird-related somehow.

**Comment by maximbaz** [comment]
Okay, this one is really important.

Some *.desktop files have a field `NoDisplay=true`. We've got to ignore such files, they are causing a lot of noise.

**Comment by maximbaz** [comment]
Actually apart from that it's looking rather good. :+1:

I wish there were app icons as well, but that's not very critical. 

For your information, to get the icon, you need to read the `Icon` field of the *.desktop file and then look for a file in `/usr/share/icons/hicolor/<resolution>/apps/<app-name>.<extension>`. At least that's how it is on my linux.

Furthermore, looks like the resolutions `32x32` and `256x256` are the most common ones, others exist not for all apps.

If I were you, I'd take a simple approach, read the value of the `Icon` field and use the `/usr/share/icons/hicolor/256x256/apps/<icon-value>.png` if it exists.

File: app/main/plugins/core/basic-apps/linux.js (hunk 2 above)
Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: Yes, that worked, just ensure the initial slash exists, and the `/usr/share/applications` (and others) are taken into account.

**Comment by maximbaz** [comment]
Another observation: I can't search using a substring. If I enter "chro" or "goog" I find "Google Chrome", but if I enter "hrome" or "oogle" I find nothing.

This is important for apps like "Qtransmission Bittorrent Client", I'd love to find it searching for "torrent".

Not sure if you want to handle this here or as a separate issue though.

**Comment by KELiON** [comment]
Ok, new test build uploaded (Linux fixes):

– [x] Do not show .desktop files with `NoDisplay=true`;
– [x] Fix for root directory `usr/local` →`/usr/local/`;
– [x] Fix exec command with `%` arguments;
– [x] Lookup for icon by `Icon=` field:

screen shot 2017-02-01 at 21 53 48

**Comment by maximbaz** [comment]
Cool, thanks a lot @KELiON !

I'm going to test this later today, I'll report with results.

**Comment by maximbaz** [comment]
I confirm all the items, nice job!

File: app/main/plugins/core/basic-apps/linux.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  ]
+}
+
+// Icon resolutions in priority of checking
+const iconResolutions = [
+  '128x128',
+  '96x96',
+  '64x64',
+  '256x256',
+  '512x512'
```
Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: out of curiosity, why not always go from high-res to low-res? why 64x64 has higher priority than 256x256?

Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: I know that you have `max-width: 200px` anyway, but to me 512px scaled down to 200px still looks nicer than 64px.

To make the point, an example of 64x64 vs 256x256:

cerebro_288

cerebro_289

Comment (KELiON) [inline app/main/plugins/core/basic-apps/linux.js]: @maximbaz maybe it makes sense. When I set this priority I thought only about productivity in results list. On MacOS I had a problem that rendering and scrolling in long result list was slow with big icons,  and after I set smaller size it became several times faster. But on MacOS these icons are base64-encoded strings, not absolute paths to files, so it also plays a role.

Would be nice to test productivity of big result list with priority from highest to lowest resolution.  I'm checking it in virtual machine so I don't have a lot of applications and vm is slower than real machine.

**Comment by vessd** [comment]
@KELiON I can't build the package and test it.
```
/home/sergey/.cache/electron-builder/AppImage/AppImage-09-07-16-linux/xorriso: error while loading shared libraries: libreadline.so.6: cannot open shared object file: No such file or directory
```
```
$ yaourt -Ql readline | rg libreadline.so
18:readline /usr/lib/libreadline.so
19:readline /usr/lib/libreadline.so.7
20:readline /usr/lib/libreadline.so.7.0
```
I will try to solve it later.

**Comment by maximbaz** [comment]
@vessd workaround: `sudo ln -s /usr/lib/libreadline.so.7 /usr/lib/libreadline.so.6`.
However if you do this, AppImage file will start showing annoying popups on every startup :wink: 

To test this branch better use [cerebro-git](https://aur.archlinux.org/packages/cerebro-git) if you are on Arch, just append `#branch=feature/windows-and-linux-apps` to the `source` in PKGBUILD.

File: app/main/plugins/core/basic-apps/linux.js (hunk 3 above)
Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: OK, I can see why on MacOS this becomes a problem if the icons have to be converted to base64 strings...

I've just set the list as below and tested on my laptop, the search still works instantly, I don't see any delay after I type a character.

```
// Icon resolutions in priority of checking
const iconResolutions = [
  '1024x1024',
  '512x512',
  '256x256',
  '192x192',
  '128x128',
  '96x96',
  '72x72',
  '64x64',
  '48x48',
  '40x40',
  '36x36',
  '32x32',
  '24x24',
  '22x22',
  '20x20',
  '16x16'
]
```

I also expanded the list with all available icon folders that I have on my box. For example, `teamviewer` ships only `48x48`, so with this list I can at least see some icon.

File: app/main/plugins/core/basic-apps/linux.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+const getId = (filePath) => {
+  const match = filePath.match(/\/applications\/(.+)$/)
+  return match ? match[1] : filePath
+}
+
+const findIcon = (icon) => {
+  if (path.isAbsolute(icon)) {
+    return icon
+  }
+  return iconDirs.map(dir => path.join(dir, `${icon}.png`)).find(fs.existsSync)
```
Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: @KELiON how difficult would it be to also support `*.svg` format in addition to `*.png`? 

SVG is small and can be scaled to any resolution, these icons should definitely have priority if they exist. I have ~20 apps that ship *.svg icons with them.

The location is `/usr/share/icons/hicolor/scalable/apps/*.svg`, so perhaps just add the top-priority resolution "scalable" to the list.

Comment (KELiON) [inline app/main/plugins/core/basic-apps/linux.js]: @maximbaz nice! added svg lookup

Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: It works, so cool!

File: app/main/plugins/core/basic-apps/linux.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  ...iconResolutions.map(resolution => (
+    appDirs.map(dir => path.join(dir, 'icons', 'hicolor', resolution, 'apps'))
+  )),
+  path.join('/usr', 'share', 'pixmaps'),
+  path.join('/usr', 'share', 'app-install', 'icons')
+])).filter(fs.existsSync)
+
+const svgIconDirs = [
+  path.join('/usr', 'share', 'icons', 'hicolor', 'scalable', 'apps')
+].filter(fs.existsSync)
```
Comment (maximbaz) [inline app/main/plugins/core/basic-apps/linux.js]: @KELiON I think you should consider reusing `iconDirs`, there you have a good set of many possible icon locations, I'm sure they cover more possibilities. I'm simply not sure that `/usr/share/icons/hicolor/scalable` is the only possible location.

Perhaps just add `scalable` as the first resolution, then `iconDirs` would visit these folders as well.

Comment (KELiON) [inline app/main/plugins/core/basic-apps/linux.js]: @maximbaz agree, it is better

**Pull request merged**